index.rst

```diff
@@ -1,29 +1,988 @@
+    This file then collects things that may eventually need to become configuration settings.
+
+``exceptions.py``
+    Any custom exceptions for this application.
+    It may be useful to define an exception parent class and then install a global handler for that exception class that generates the correct HTTP error code and body structure.
+    Then, all handlers and even services can raise that exception without catching it, and the code to translate it into a valid HTTP error reply can be shared.
+    Good candidates for this are a ``ValidationError`` that generates a 422 error compatible with FastAPI and a ``PermissionDeniedError`` that generates a 403 error.
+
+    Exception class names should generally end in ``Error`` (not ``Exception``) following :pep:`8`.
+
+    It's often a good idea to define custom constructors for exceptions that take specific, well-defined, typed data and then construct the human-readable message in the exception code, for better code sharing.
+
+    For exceptions designed to generate structured JSON bodies as part of HTTP errors, define a ``to_dict`` method that translates the exceptions into a dictionary suitable for serializing to JSON.
+
+``factory.py``
+    Contains the factory object used to construct services and their dependencies.
+    Use of the factory pattern is optional and may not be appropriate for smaller applications.
+
+``main.py``
+    Defines the FastAPI application.
+    This should either create a global variable named ``app`` or a function named ``create_app``, depending on whether all application initialization can be done at module load time.
+    The main case where a ``create_app`` function may be required is if the application object depends on the configuration and the configuration is loaded from a YAML file (see :ref:`configuration`).
+    Using a function then allows delaying loading the configuration until a test case has a chance to switch to a different configuration file than the default.
+
+    This module should register all of the routers, set up any middleware, set up any exception handlers, and handle startup and shutdown events.
+    Exception handlers can be defined in this same module unless they are complex (they normally won't be).
+    The startup and shutdown handlers are conventionally named ``startup_event`` and ``shutdown_event``, respectively, and should handle initializing and closing any dependencies that hold state or external connections.
+
+``util.py``
+    Random utility functions used by the rest of the code.
+    This should only contain simple functions and should not contain any business logic.
+    All business logic should go into a service object instead.
+    This is a good place to put Pydantic validators that are shared by multiple models.
+
+If this application uses a SQL database for storage, the SQLAlchemy_ ORM models should go into a directory named ``schema``, and the ``__init__.py`` file for that directory should import all of the models.
```

Comment:
In some cases I've renamed this to "dbschema" because "schema" can get a bit overloaded.